src/room_member.cpp

```diff
@@ -154,6 +154,37 @@ void RoomMember::SendWifiPacket(const WifiPacket& wifi_packet) {
     peer->Send(&stream, HIGH_PRIORITY, RELIABLE, 0, server_address, false);
 }
 
+void RoomMember::Register(Callback callback, EventType event_type) {
+    std::lock_guard<std::mutex> lock(callback_mutex);
+
+    auto callbacks = atomic_load(&callback_map.at(event_type));
+    auto new_callbacks = std::make_shared< std::vector<Callback> >();
+    new_callbacks->reserve(callbacks->size()+1);
+    *new_callbacks = callbacks;
+    new_callbacks->push_back(std::move(callback);
+
+    atomic_store(&callback_map.at(event_type), new_callbacks);
+}
+
+void RoomMember::Deregister(Callback callback, EventType event_type) {
+    std::lock_guard<std::mutex> lock(callback_mutex);
+
+    auto callbacks = atomic_load(&callback_map.at(event_type));
+    auto new_callbacks = std::make_shared< std::vector<Callback> >();
+    new_callbacks->reserve(callbacks->size());
+    *new_callbacks = callbacks;
+    auto itr = std::find(new_callbacks->begin(), new_callbacks->end(), callback);
+    new_callbacks->erase(itr);
+
+    atomic_store(&callback_map.at(event_type), new_callbacks);
+}
+
+void RoomMember::Invoke(EventType event_type)
+{
+    auto callbacks = atomic_load(&callback_map.at(event_type));
+    for(auto& callback: *callbacks) (*callback)();
```

Comment:
Please always make this more explicit - at least in 2 lines, I still prefer explicit scopes (but apparently we don't do the extra scope in Citra [WHYYY?!! goto fail?!?!?!!!])